src/network/networkd-ndisc.c

```diff
@@ -897,21 +900,47 @@ static int ndisc_router_process_captive_portal(Link *link, sd_ndisc_router *rt)
                 return log_link_warning_errno(link, r, "Failed to get captive portal from RA: %m");
 
         if (len == 0)
-                return 0;
+                return log_link_warning_errno(link, SYNTHETIC_ERRNO(EBADMSG), "Received empty captive portal, ignoring.");
 
         r = make_cstring(uri, len, MAKE_CSTRING_REFUSE_TRAILING_NUL, &captive_portal);
         if (r < 0)
                 return log_link_warning_errno(link, r, "Failed to convert captive portal URI: %m");
 
         if (!in_charset(captive_portal, URI_VALID))
-                return log_link_debug_errno(link, SYNTHETIC_ERRNO(EBADMSG), "Received invalid captive portal, ignoring.");
+                return log_link_warning_errno(link, SYNTHETIC_ERRNO(EBADMSG), "Received invalid captive portal, ignoring.");
+
+        if (lifetime_usec == 0) {
```

Comment:
I wonder if this is reachable now that https://github.com/systemd/systemd/commit/4ccd48cb2693a3603f14a30b6286228597285d92 is merged? It seems to me that this PR is out of sync with that commit.

Reply:
Indeed. Good catch.

Reply:
FWIW looks like 4ccd48cb2693a3603f14a30b6286228597285d92 is buggy because it prevents DNSSLs and RDNSSs with zero lifetimes from being removed if they are initially advertised with the "infinity" lifetime and reset in RAs with zero lifetimes. I think it would be better to fix that weird logic first because I have to admit I'm not sure how this stuff works.

Reply:
Hm? I cannot follow. The bug you mean is #28976?

Anyway, that's out of scope of this PR. Please open a new issue page if #28976 is not related what you want to say.

Reply:
> The bug you mean is https://github.com/systemd/systemd/pull/28976?

As far as I can tell with that PR applied DNSSLs and RDNSSs with their own zero lifetimes are still ignored if they come in RAs with zero lifetimes because `ndisc_router_process_options` is just skipped but yeah those `if` statements are problematic.

> that's out of scope of this PR

Agreed but I didn't mean to say that it should be fixed here. My point is that the zero lifetime logic is messed up and since everything else is built on top of it it should be fixed first.